### PR TITLE
Prime number challenge

### PR DESCRIPTION
**Pull request by doctorbanu**
---
## about: A function that checks whether a given non-negative integer is a prime number.
---

## Behavior
This function checks if a given integer n is a prime number. A prime number is a number greater than 1 that has no divisors other than 1 and itself.
* Input:
The function accepts a single integer n, where n must be a non-negative integer (i.e., n >= 0).
* Output:
Returns True if n is a prime number.
Returns False if n is not a prime number.
* Error Handling:
If the input is not an integer or if n is negative, the function raises an AssertionError.
* Calculation:
If n <= 1, return False (since prime numbers must be greater than 1).
Check divisibility of n from 2 up to the square root of n. If n is divisible by any of these numbers, return False.
If no divisors are found, return True.

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/prime_number_checker.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Prime Number Checker
+=============================
+This script provides a function to check if a number is a prime number or not.
+
```
Comment (misik-eng) [inline solutions/prime_number_checker.py:7]: Your module docstring looks great! Consider enhancing it by including details about the input, such as a note specifying that it should be a non-negative integer. Additionally, are there other constraints? For instance, what’s the upper limit for the input? I assume it’s not infinite! :)

File: solutions/prime_number_checker.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+Module contents:
+    - is_prime: Checks if a given number is a prime number.
+
+Author: Banu Ozyilmaz
+Created on: 12-29-2024
+"""
+
+
+def is_prime(n: int) -> bool:
```
Comment (misik-eng) [inline solutions/prime_number_checker.py:16]: Does the function's name matches the file name?  It’s descriptive and looks good, but perhaps we could choose more similar names for  file and function?
 !

File: solutions/prime_number_checker.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        False
+        >>> is_prime(2)
+        True
+        >>> is_prime(3)
+        True
+        >>> is_prime(4)
+        False
+        >>> is_prime(17)
+        True
+        >>> is_prime(18)
```
Comment (misik-eng) [inline solutions/prime_number_checker.py:41]: It seems you have enough doctests, but perhaps we could replace one of the small-number tests with a larger number?

File: solutions/prime_number_checker.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    assert isinstance(n, int), "Input must be an integer."
+    assert n >= 0, "Input must be a non-negative integer."
+
+    # Prime numbers are greater than 1
+    if n <= 1:
+        return False
+
+    # Check for factors from 2 up to the square root of n
+    # If n has a factor smaller than its square root, it must also have a larger one,
+    # therefore n can't be prime.
```
Comment (misik-eng) [inline solutions/prime_number_checker.py:62]: Do we need that long comment?  or we can split longer line into half , like:
```
# If n has a factor smaller than its square root,
#it must also have a larger one, therefore n can't be prime.
```

File: solutions/prime_number_checker.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    for i in range(2, int(n**0.5) + 1):
```
Comment (misik-eng) [inline solutions/prime_number_checker.py:63]: It is shorter with your solution but maybe We can use a variable to be more easy to understand and have more readable for loop, what do you think?
 like:

```
sqrt_limit = int(n**0.5) + 1
for i in range(2, sqrt_limit):
```

**Review by misik-eng — DISMISSED** [review]
Looks almost perfect! There are just a few small things to reconsider, but I approve this merge. Thank you!
I’ve used our checklist, and it looks great!
I won't merge it now, because maybe you might want to make some changes before we merge.

### Files

- [x] **The file name describes the function's behavior**


> Does the function's name match the file name? It’s descriptive and looks good, but perhaps we could choose more similar names for the file and function.

- [x] There is a module docstring in the function file

> Your module docstring looks great! Consider enhancing it by including details about the input, such as a note specifying that it should be a non-negative integer. Additionally, are there other constraints? For instance, what’s the upper limit for the input? I assume it’s not infinite!**

- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

> It seems you have enough doctests, but perhaps we could replace one of the small-number tests with a larger number?

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

## Strategy

### Do's

- [x] Variable names help to understand the strategy

> > It is shorter with your solution, but maybe we can use a variable to make it easier to understand and have a more readable "for" loop. For example:
> ```
> sqrt_limit = int(n**0.5) + 1  
> for i in range(2, sqrt_limit):  
> ```

- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation
 

> >  Do we need that long comment? Perhaps we can split it into two shorter lines for better readability:
> ```
> # If n has a factor smaller than its square root,  
> # it must also have a larger one. Therefore, n can't be prime. 
> ```

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

**Review by misik-eng — APPROVED** [review]

File: solutions/prime_number_checker.py (hunk 1 above)
Comment (doctorbanu) [reply solutions/prime_number_checker.py:7]: The function handles up to 9223372036854775807 on 64-bit systems; beyond that, it's memory-dependent. I don't think it is necessary for this case.

File: solutions/prime_number_checker.py (hunk 2 above)
Comment (doctorbanu) [reply solutions/prime_number_checker.py:16]: yes you are right. I have changed the file name.

File: solutions/prime_number_checker.py (hunk 3 above)
Comment (doctorbanu) [reply solutions/prime_number_checker.py:41]: I have added larger number. Thank you

File: solutions/prime_number_checker.py (hunk 5 above)
Comment (doctorbanu) [reply solutions/prime_number_checker.py:63]: Thank you, you are right. I thought it would be easier on the eyes in a single line, but this step makes it more understandable. Thanks, I have corrected it this way.

**Review by misik-eng — APPROVED** [review]
This time,  looks like there is no conflicts or duplicated files:) perfect job, awesome!!

**Comment by doctorbanu** [comment]
No, I thank YOU so much. You were a great guide and helped me see and fix my mistakes. Your guidance really helped me overcome the problems. I've learned a lot thanks to you, and you always showed me the right way when I needed help. Thanks again Muhammet, I really appreciate your help.
